packages/core/server/internal-base/BUILD.bazel

```diff
@@ -0,0 +1,98 @@
+load("@npm//@bazel/typescript:index.bzl", "ts_config")
+load("@build_bazel_rules_nodejs//:index.bzl", "js_library")
+load("//src/dev/bazel:index.bzl", "jsts_transpiler", "pkg_npm", "pkg_npm_types", "ts_project")
+
+PKG_DIRNAME = "internal-base"
+PKG_REQUIRE_NAME = "@kbn/core-server-internal-base"
```

Comment:
Yea, could work too. I find it more explicit to separate `common|public|server` at a higher folder level that the service name (and also, some packages will not directly be associated to a given service). Curious to see what @elastic/kibana-operations thinks about it

Reply:
After thinking about this lately along with @spalger we think a good standard for this is `shared-common|shared-browser|shared-server` and this is what we are going to purpose in a RFC under work at the moment (we also have other types but not important here from what I can see). Having that said I think then the project arrangement of `packages/core/{service-name}/(shared-common|shared-browser|shared-server)` is what better serves the purpose here as it follows the npm idioms and also organizes the code by domain resulting. \cc @spalger

Reply:
Yeah, we've been using `common`/`browser`/`server` for the three big "targets". Feels like a reasonable time to move from `public` to `browser`. I don't think we need to explicitly mention `shared-` in directories, that's more of a way to differentiate package types but I think we can assume things are shared by default so I personally would advocate for `packages/core/{service-name}/(common|browser|server)` but I also think whatever the core team likes best would be good.

Reply:
> the project arrangement of packages/core/{service-name}/(shared-common|shared-browser|shared-server) is what better serves the purpose here

I get the `common|browser|server` part, but what's the `shared` part of it? is it for `shared` versus `internal`?

Reply:
No, it's for shared vs plugin vs functional test packages which can't be imported or depended on. They're not "internal" necessarily but they're not for shared code.

Reply:
> I personally would advocate for packages/core/{service-name}/(common|browser|server)

I think it make sense to have the functional domain being at a higher level than `common|browser|server`. We even discussed this structure a few years back with the team (when we were still called `platform` 😅 ), and agreed that this was were we wanted to go long-term, so overall it's probably a great opportunity to do it.

However, we have some 'service domains' that will have more than one package, and I wonder how exactly the structure should look like in such cases.

To take a concrete example, the `elasticsearch` domain will have at least two 'groups' of packages on the server-side: one for the service (atm the `src/core/server/elasticsearch` folder), and one for the ES client wrapper (atm the `src/core/server/elasticsearch/client` folder).

Should the structure then be something like:

```
- packages
  - core
      - elasticsearch
        - server
          - public-types   // @kbn/core-elasticsearch-server-public-types
          - service        // @kbn/core-elasticsearch-server-service
          - mocks          // @kbn/core-elasticsearch-server-mocks
          - client-types   // @kbn/core-elasticsearch-server-client-types
          - client         // @kbn/core-elasticsearch-server-client 
          - client-mocks   // @kbn/core-elasticsearch-server-client-mocks
```

Or do we prefer to isolate the 'sub' packages of a service domain at a higher level, e.g

```
- packages
  - core
      - elasticsearch
        - server
          - public-types   // @kbn/core-elasticsearch-server-public-types
          - service        // @kbn/core-elasticsearch-server-service
          - mocks          // @kbn/core-elasticsearch-server-mocks
      - elasticsearch-client
        - server 
          - public-types   // @kbn/core-elasticsearch-client-server-public-types
          - client         // @kbn/core-elasticsearch-client-server
          - mocks          // @kbn/core-elasticsearch-client-server-mocks
```

Or maybe we want to group by sub-domain after the `server` folder?

Or do we prefer to isolate the 'sub' packages of a service domain at a higher level, e.g

```
- packages
  - core
      - elasticsearch
        - server
          - service 
            - public-types   // @kbn/core-elasticsearch-server-service-public-types
            - service        // @kbn/core-elasticsearch-server-service
            - mocks          // @kbn/core-elasticsearch-server-service-mocks
          - client
            - public-types   // @kbn/core-elasticsearch-server-client-public-types
            - client         // @kbn/core-elasticsearch-server-client
            - mocks          // @kbn/core-elasticsearch-server-client-mocks
```

Writing it, I think I find the first option a bit too 'messy', and I kinda prefer the third option over the second one, as it makes a better distinction between the sub-domains while keeping everything under the same domain (`elasticsearch`) folder. The drawback is that for domains that will only have one 'package group' (which is, most of our domains), we will be creating an additional level of folder unnecessarily (e.g `packages/core/server/docLinks/service` instead of just `packages/core/server/docLinks`).

Curious to see what @elastic/kibana-operations and @elastic/kibana-core think about it.

Also, I have to admit, even if that doesn't follow the folder structure, I kinda prefer having the `(common|browser|server)` prefix in the package name **before** the service. It feels more... natural to me maybe? e.g `@kbn/core-server-elasticsearch-public-types` instead of `@kbn/core-elasticsearch-server-public-types`. But that may be only resistance to change because that go the opposite way of how we're 'thinking' when importing from Core atm. WDYT on that point too?

In term of folder/package naming, I'm open to alternatives to the proposed suffixes, in particular for `public-types`, but as discussed with @spalger yesterday, I do think we need a easy way to distinguish which core packages can be imported from plugins, so we need a convention for package names regarding which ones can be imported outside of core.

Reply:
Let me first highlight that I don't have a strong opinion. I'd just like to ask a question:

How are we going to import these packages? I think answering this question will help us find out the best structure.

For example, at the moment, we structure core's directories as `src/core/(public|server)`, and then all the core services sit inside those directories.

However, any other services outside core follow the `{serviceName}/(common|public|server)` structure. Only that we call them _plugins_: `(src|x-pack)/plugins/{serviceName}/(common|public|server)`

What's the vision with the packages? Will it be allowed for other services outside core to directly import the `i18n` service (only to name one)? If that's the case, I think we should aim for having the service name closer to the top to make a clear distinction between `@kbn/*i18n-common`, `@kbn/*i18n-browser` and `@kbn/*i18n-server`.

If the intention is to follow our current core's single entry point approach, but from a package, I'd say it's more an implementation detail, and it _makes sense_ to have `packages/core/(common|public|server)/{coreServiceName})` because external services will only import from `@kbn/core-(common|public|server)`.

What do you think?

Reply:
> What's the vision with the packages? Will it be allowed for other services outside core to directly import the i18n service (only to name one)?

The end goal is to get rid of the current entrypoints (`src/core/{server,public}/index.ts`) have every plugin imports core types directly from core packages. However, they will not be allowed to import from 'private' core packages. For example they'll be able to import the i18n service contracts from the `public-type` package of i18n, but an eslint rule will block them from importing from the `service` package, that is internal to core (note: names/suffixes are all but final here)

Reply:
Understood, this is a matter of taste, to be fair, but I guess it feels more natural for the package to be named `@kbn/core-i18n-(common|public|server)-public-types`?

Now my next question? Can we make it somehow that the eslint rule only allows `import type` of the services but not plain `import`s? Would that save us from splitting the `service` and `public-type` packages?

Reply:
> Can we make it somehow that the eslint rule only allows import type of the services but not plain imports?

Not sure, but that wouldn't be good enough. Like, we don't want external importers to be able to import `InternalHttpSetup` for instance.

Now, to be honest, we could also just decided that it's okay for external consumers to import anything from Core. It's just that it's not the case at the moment, and this was done for a reason. Ideally, we want to preserve the same import restrictions. 

Also, it has the side-effect upside of forcing us to solve all our internal/cross dependency issues, which will be time consuming, but in the end is probably a very beneficial thing.

Reply:
My two cents: I think dropping the `-public-types` naming and just having `@kbn/core-(service)-(common|browser|server)` be the public types, then have an additional `@kbn/core-i18n-browser-impl` or `-internal` package for the implementation helps with understanding and makes it more clear to people what they should and shouldn't import.

Reply:
Yes, from an API consumer point of view, I agree that having a specific 'public' suffix probably doesn't make sense, and that we should instead suffix our 'internal' packages only. I prefer the `-internal` suffix rather than `-impl` though, as we may have more than just 'implementation' internal packages. 

E.g for the server-side elasticsearch service's packages:
```
// public package (should only contain types in theory)
@kbn/core-elasticsearch-server
// internal types and concrete implementation 
@kbn/core-elasticsearch-server-internal 
// public (and probably internal too, don't think its worth it splitting the two) mocks
@kbn/core-elasticsearch-server-mocks 
```

The only remaining question is about the folder structure then. Need to think a bit more